gui/biomes.lua

```diff
@@ -0,0 +1,213 @@
+                table.insert(biomeList, {biome = biome, char = ch, typeId = biomeTypeId})
+                ix = #biomeList
+            end
+
+            biomesZY[x] = ix
+        end
+    end
+end
+
+gatherBiomeInfo()
+
+local TITLE = "Biomes"
+
+BiomeVisualizerLegend = nil -- for debugging purposes
+BiomeVisualizerLegend = defclass(BiomeVisualizerLegend, widgets.Window)
+BiomeVisualizerLegend.ATTRS {
+    frame_title=TITLE,
+    frame_inset=0,
+    resizable=true,
+    autoarrange_subviews = true,
+    autoarrange_gap = 1,
+    frame = {
+        w = 47,
+        h = 10,
+        l = 5,
+        t = 8,
+    },
+}
+
+local function GetBiomeName(biome, biomeTypeId)
+    -- based on probe.cpp
+    local sav = biome.savagery
+    local evi = biome.evilness;
+    local sindex = sav > 65 and 2 or sav < 33 and 0 or 1
+    local eindex = evi > 65 and 2 or evi < 33 and 0 or 1
+    local surr = sindex + eindex * 3 +1; --in Lua arrays are 1-based
+
+    local surroundings = {
+        "Serene", "Mirthful", "Joyous Wilds",
+        "Calm", "Wilderness", "Untamed Wilds",
+        "Sinister", "Haunted", "Terrifying"
+    }
+    local surrounding = surroundings[surr]
+
+    local name = biomeTypeNames[df.biome_type[biomeTypeId]] or "DFHACK_Unknown"
+
+    return ([[%s %s]]):format(surrounding, name)
+end
+
+function BiomeVisualizerLegend:init()
+    self:addviews{
+        widgets.List{
+            view_id = 'list',
+            frame = { t = 0 },
+            text_pen = { fg = COLOR_GREY, bg = COLOR_BLACK },
+            --cursor_pen = { fg = COLOR_BLACK, bg = COLOR_GREEN },
+            --on_select = self:callback('onSelectEntry'),
+        },
+    }
+
+    local choices = {}
+    for _, biomeExt in ipairs(biomeList) do
+        table.insert(choices, {
+            text = ([[%s: %s]]):format(biomeExt.char, GetBiomeName(biomeExt.biome, biomeExt.typeId)),
+            biomeTypeId = biomeExt.typeId,
+        })
+    end
+    self.subviews.list:setChoices(choices)
+end
+
+BiomeVisualizer = defclass(BiomeVisualizer, gui.ZScreen)
+BiomeVisualizer.ATTRS{
+    focus_path='BiomeVisualizer',
+}
+
+local function getMapViewBounds()
+    local dims = dfhack.gui.getDwarfmodeViewDims()
+    local x = df.global.window_x
+    local y = df.global.window_y
+    return {x1 = dims.map_x1 + x,
+            y1 = dims.map_y1 + y,
+            x2 = dims.map_x2 + x,
+            y2 = dims.map_y2 + y,
+    }
+end
+
+function BiomeVisualizer:init()
+    self:addviews{BiomeVisualizerLegend{}}
+end
+
+function BiomeVisualizer:onRenderFrame(dc, rect)
+    BiomeVisualizer.super.onRenderFrame()
+
+    local function get_overlay_pen(pos)
+        local safe_index = safe_index
+        -- 304 = `0`, 353 = `a`
+        local idxBaseTile = 353
+        local biomes = biomesMap
+
+        local N = safe_index(biomes, pos.z, pos.y, pos.x)
+        if not N then return end
+        return COLOR_RED, tostring(N), idxBaseTile + (N - 1)
+    end
+    local overlay_bounds = getMapViewBounds()
```

Comment:
if you're iterating over the entire visible viewport, you can pass in `nil` for `overlay_bounds`